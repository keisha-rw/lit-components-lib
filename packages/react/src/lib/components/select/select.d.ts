import * as React from 'react';
export declare const PdsSelect: React.FunctionComponent<Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
