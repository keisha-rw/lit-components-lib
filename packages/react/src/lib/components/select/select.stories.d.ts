import * as React from 'react';
declare const _default: {
    title: string;
    component: React.FunctionComponent<Partial<{
        size: "default" | "sm";
        label: string | undefined;
        disabled: boolean;
        value: string;
        hideLabel: boolean;
        fieldId: string | undefined;
        required: boolean;
        readonly: boolean;
        helpText: string | undefined;
        errorMessage: string;
        placeholder: string | undefined;
    }> & {
        ref?: React.Ref<any> | undefined;
    } & {
        style?: React.CSSProperties | undefined;
    } & Required<{
        name: string;
    }> & Partial<{
        onChange: (e: Event) => void;
        onBlur: (e: Event) => void;
        onFocus: (e: Event) => void;
    }> & {
        id?: string | undefined;
        slot?: string | undefined;
        className?: string | undefined;
        children?: React.ReactNode;
    }>;
    tags: string[];
};
export default _default;
export declare const Default: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const DefaultSmall: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const Disabled: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const DisabledWithHelpText: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const Error: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const Placeholder: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const DefaultValue: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const HelpText: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const HiddenLabel: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    size: "default" | "sm";
    label: string | undefined;
    disabled: boolean;
    value: string;
    hideLabel: boolean;
    fieldId: string | undefined;
    required: boolean;
    readonly: boolean;
    helpText: string | undefined;
    errorMessage: string;
    placeholder: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Required<{
    name: string;
}> & Partial<{
    onChange: (e: Event) => void;
    onBlur: (e: Event) => void;
    onFocus: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
