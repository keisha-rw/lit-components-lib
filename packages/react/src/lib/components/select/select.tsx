import { EventName } from '@lit-labs/react';
import { PdsSelect as PdsSelectElement } from '@keisha/design-system';
import * as React from 'react';
import { PdsSelectProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsSelect = createPdsReactComponent<PdsSelectProps>()({
  tagName: 'pds-select',
  elementClass: PdsSelectElement,
  react: React,
  events: {
    onChange: 'pds-select-change' as EventName<CustomEvent>,
    onBlur: 'pds-select-blur' as EventName<CustomEvent>,
    onFocus: 'pds-select-focus' as EventName<CustomEvent>,
  },
});
