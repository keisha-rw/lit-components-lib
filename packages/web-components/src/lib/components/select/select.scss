@import '../../../../../tokens/tokens/01-base/global-utils/a11y';

/*------------------------------------*\
 #SELECT_MENU
\*------------------------------------*/

.pds-c-select__select {
  @include pds-semantic-body-default;

  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;
  padding: size(0.75) size(6.5) size(0.75) size(2);
  gap: size(2);
  isolation: isolate;
  box-shadow: none;
  width: 100%;
  height: size(6);
  appearance: none;
  border-radius: size(0.5);
  color: var(--pds-semantic-text-default);
  background-color: var(--pds-semantic-background-default);
  border: size(0.125) solid var(--pds-semantic-border-strong);

  &:hover {
    background-color: var(--pds-semantic-background-subtle);
    border: size(0.125) solid var(--pds-semantic-border-interactive-strong);
  }

  &:focus-visible {
    outline: none;
  }

  &:focus {
    background-color: var(--pds-semantic-background-subtle);
    border: size(0.125) solid var(--pds-semantic-border-interactive-xstrong);
    box-shadow: size(0) size(0) size(1)
      var(--pds-semantic-border-interactive-xstrong);
  }

  .pds-c-select--sm & {
    height: size(5);
  }

  .pds-is-error & {
    border-color: var(--pds-semantic-border-error);

    &:focus {
      box-shadow: 0 0 size(0.5) var(--pds-semantic-border-error);
    }
  }

  /* stylelint-disable-next-line no-descending-specificity */
  .pds-is-disabled & {
    cursor: not-allowed;
    background: var(--pds-semantic-background-disabled);
    border: size(0.125) solid var(--pds-semantic-border-disabled);

    &:hover {
      border: size(0.125) solid var(--pds-semantic-border-disabled);
    }
  }
}

.pds-c-select__label {
  @include pds-semantic-form-label-default;
  // Inside auto layout
  flex: none;
  order: 0;
  align-self: stretch;
  flex-grow: 0;

  .pds-is-disabled & {
    color: var(--pds-semantic-text-disabled);
    cursor: not-allowed;
  }

  .pds-c-select--hidden-label & {
    @include pds-util-sr-only;
  }
}

.pds-c-select__help-text {
  @include pds-semantic-body-sm;

  color: var(--pds-semantic-text-default);
  line-height: size(2.25);
  margin-bottom: size(1);

  .pds-is-disabled & {
    cursor: not-allowed;
    color: var(--pds-semantic-text-disabled);
  }
}

.pds-c-select__required-indicator {
  color: var(--pds-semantic-text-error);
}

.pds-c-select__error-message {
  @include pds-semantic-body-sm;

  color: var(--pds-semantic-text-error);
  display: flex;
  align-items: center;
  gap: size(1);
  margin-top: size(0.25);
}

.pds-c-select__down {
  position: absolute;
  right: size(1.5);
  pointer-events: none;
  color: var(--pds-semantic-border-icon-default);

  .pds-is-disabled & {
    color: var(--pds-semantic-border-disabled);
  }
}

.pds-c-select__wrapper {
  position: relative;
  display: flex;
  align-items: center;
}

// options are originally slotted, but then copied inside the component
// so using a element selector is necessary here
option:checked {
  @include pds-semantic-body-bold-default;
}
