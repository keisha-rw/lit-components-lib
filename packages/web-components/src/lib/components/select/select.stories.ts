import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './select';

export default {
  title: 'Components/Select',
  component: 'pds-select',
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: 'radio',
      options: ['default', 'sm'],
    },
    value: {
      control: 'text',
    },
    fieldId: {
      control: 'text',
    },
    label: {
      table: {
        category: 'attributes',
      },
      description:
        ' The label of the form field. Must be plain text. If label requires additional markup (e.g., superscript), then use the label slot instead.',
      control: 'text',
    },
    'slot:label': {
      name: 'label',
      control: 'text',
      table: {
        category: 'slots',
      },
      description:
        'Use this slot instead of the label property, if the label requires additional markup.',
    },
    helpText: {
      control: 'text',
    },
    errorMessage: {
      control: 'text',
    },
    readonly: {
      table: {
        disable: true,
      },
    },
  },
  parameters: {
    actions: {
      handles: ['pds-select-change'],
    },
  },
  render: (args) =>
    html`<pds-select
      label=${args.label || nothing}
      size=${args.size || nothing}
      name="select-field"
      ?disabled=${args.disabled}
      ?required=${args.required}
      ?hideLabel=${args.hideLabel}
      helpText=${args.helpText || nothing}
      placeholder=${args.placeholder || nothing}
      errorMessage=${args.errorMessage || nothing}
      value=${args.value || nothing}
    >
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
    </pds-select>`,
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963
export const Default: StoryObj = {
  name: 'Default',
  args: {
    size: 'default',
    label: 'This is a label',
    placeholder: '',
  },
};

export const DefaultSmall: StoryObj = {
  name: 'Default small',
  args: {
    size: 'sm',
    label: 'This is a label',
    placeholder: '',
  },
};

export const Disabled: StoryObj = {
  name: 'Disabled',
  args: {
    disabled: true,
    label: 'This is a label',
  },
};

export const DisabledWithHelpText: StoryObj = {
  name: 'Disabled with help text',
  args: {
    disabled: true,
    helpText:
      'Here is some more text explaining why you should choose an option',
    label: 'Choose an option',
  },
  parameters: {
    a11y: {
      config: {
        rules: [
          {
            // This is a disabled field, so we can ignore this check
            //
            // @link https://dequeuniversity.com/rules/axe/4.6/color-contrast
            id: 'color-contrast',
            enabled: false,
          },
        ],
      },
    },
  },
};

export const Error: StoryObj = {
  name: 'Error',
  args: {
    required: true,
    label: 'This is a label',
    placeholder: '',
    errorMessage: 'Error message goes here.',
  },
};

export const Placeholder: StoryObj = {
  name: 'Placeholder',
  args: {
    required: true,
    label: 'This is a label',
    placeholder: 'Please choose an option.',
  },
};

export const PreselectedOption: StoryObj = {
  name: 'Pre-selected option',
  args: {
    required: true,
    label: 'This is a label',
    value: '2',
  },
};

export const HelpText: StoryObj = {
  name: 'Help text',
  args: {
    helpText:
      'Here is some more text explaining why you should choose an option',
    label: 'Choose an option',
  },
};

export const HiddenLabel: StoryObj = {
  name: 'Hidden label',
  args: {
    label: 'You cannot see this label, but we want it for screen readers',
    hideLabel: true,
    placeholder: 'Choose an option',
  },
};
