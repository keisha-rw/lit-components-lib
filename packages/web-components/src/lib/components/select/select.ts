import '@keisha/design-system-icons-web/chevron-down';
import { html, nothing } from 'lit';
import { unsafeHTML } from 'lit/directives/unsafe-html.js';
import { property, query, state } from 'lit/decorators.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsFormElement } from '../pds-form-element/PdsFormElement';
import styles from './select.scss?inline';

/**
 * @summary This component renders a styled select input component.
 *
 * @slot label Use this slot instead of the label property, if the label requires additional markup.
 * @slot label-after Use this slot if markup should be inserted after the label.
 * @slot help-text Use this slot instead of the helpText property, if the help text requires additonal markup.
 */
@customElement('pds-select', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsSelect extends PdsFormElement {
  /**
   * Size
   * -**sm** renders a the small version of the Select
   */
  @property()
  size: 'sm' | 'default' = 'default';

  @property({ type: String })
  placeholder?: string;

  /** @internal */
  @state()
  childNodeObserver = new MutationObserver((mutationList) => {
    const me = this;
    mutationList.forEach((mutation) => {
      const mutationTargetHtmlElement = mutation.target as HTMLElement;
      const mutationAddedNode = mutation.addedNodes[0] as HTMLElement;
      const mutationRemovedNode = mutation.removedNodes[0] as HTMLElement;

      if (
        mutationTargetHtmlElement &&
        mutationTargetHtmlElement.tagName &&
        mutationTargetHtmlElement.tagName.toLowerCase() === 'option'
      ) {
        // option attribute modified
        this.childNodeObserverCallback(me);
      } else if (
        mutationTargetHtmlElement &&
        mutationTargetHtmlElement.parentElement &&
        mutationTargetHtmlElement.parentElement.tagName &&
        mutationTargetHtmlElement.parentElement.tagName.toLowerCase() ===
          'option'
      ) {
        // option text modified
        this.childNodeObserverCallback(me);
      } else if (
        mutation.addedNodes &&
        mutationAddedNode &&
        mutationAddedNode.tagName &&
        mutationAddedNode.tagName.toLowerCase() === 'option'
      ) {
        // new option added
        this.childNodeObserverCallback(me);
      } else if (
        mutation.removedNodes &&
        mutationRemovedNode &&
        mutationRemovedNode.tagName &&
        mutationRemovedNode.tagName.toLowerCase() === 'option'
      ) {
        // option removed
        this.childNodeObserverCallback(me);
      }
    });
  });

  @query('select')
  field: HTMLSelectElement;

  /**
   * Set the selected option based on the current value
   */
  protected override updateField() {
    // for loop for the case where a browser doesn't make a collection iterable
    for (let i = 0; i < this.field?.options?.length; i += 1) {
      const option = this.field.options[i];

      if (option.value === (this.internalValue || '')) {
        option.selected = true;
        break;
      }
    }
  }

  protected override firstUpdated() {
    super.firstUpdated();

    this.updateField();
    // Options for the observer (which mutations to observe)
    const config = {
      childList: true,
      subtree: true,
      attributes: true,
      characterData: true,
    };
    this.childNodeObserver.observe(this, config);
  }

  private handleBlur() {
    this.dispatchEvent(
      new CustomEvent('pds-select-blur', {
        bubbles: false,
        composed: true,
        detail: {
          summary: this.value,
        },
      }),
    );
  }

  private handleChange() {
    this.value = this.field.value;
    this.dispatchEvent(
      new CustomEvent('pds-select-change', {
        bubbles: true,
        composed: true,
        detail: {
          summary: this.value,
        },
      }),
    );
  }

  private handleFocus() {
    this.dispatchEvent(
      new Event('pds-select-focus', { bubbles: false, composed: true }),
    );
  }

  /**
   * Check if an option exists in the select
   */
  selectContainsOption(select: HTMLSelectElement, value: string) {
    return (
      Array.from(select.options).filter((option) => option.value === value)
        .length > 0
    );
  }

  async childNodeObserverCallback(currentSelect: PdsSelect) {
    // When child options change, we need to request update of the component
    currentSelect.requestUpdate();
    await currentSelect.updateComplete;

    // check if the value still is a valid option in the select and set it
    if (this.selectContainsOption(currentSelect.field, currentSelect.value)) {
      const selectValue = currentSelect.value;
      // eslint-disable-next-line no-param-reassign
      currentSelect.value = selectValue;
      currentSelect.updateField();
    } else {
      currentSelect.formResetCallback();
      currentSelect.handleChange();
    }
  }

  /** @internal */
  get classNames() {
    return {
      [this.size]: !!this.size,
      'is-error': !!this.errorMessage,
      'is-disabled': this.disabled,
      'hidden-label': this.hideLabel,
    };
  }

  /**
   * Creates an HTML template for placeholder,
   * if the `placeholder` prop is defined.
   * Adds a hidden and disabled attribute
   * when this select is required.
   *
   * @returns an HTML template for placeholder or nothing
   */
  protected placeholderTemplate() {
    // Set prettier ignore to not enforce extra white space around the placeholder text.
    // prettier-ignore
    return this.placeholder
      ? html`<option id="placeholder"
          value=""
          ?hidden=${this.required}
          ?disabled=${this.required}
        >${this.placeholder}</option>`
      : nothing;
  }

  render() {
    if (!this.verifyLabel()) {
      return nothing;
    }

    return html`<div class="${this.getClass()}">
      ${this.labelTemplate()} ${this.helpTextTemplate()}
      <div class="${this.classEl('wrapper')}">
        <select
          class="${this.classEl('select')}"
          id="${this.fieldId || `${this.name}-field`}"
          ?disabled=${this.disabled}
          ?required=${this.required}
          aria-describedby=${this.getAriaDescribedBy() || nothing}
          aria-invalid=${this.errorMessage ? 'true' : nothing}
          name="${this.name}"
          @change=${this.handleChange}
          @focus=${this.handleFocus}
          @blur=${this.handleBlur}
        >
          ${this.placeholderTemplate()}
          ${Array.from(this.children).map((child: Element) => {
            return html`${unsafeHTML(child.outerHTML)}`;
          })}
        </select>
        <span class="${this.classEl('down')}">
          <pds-icon-chevron-down size="default"></pds-icon-chevron-down>
        </span>
      </div>
      ${this.errorMessageTemplate()}
    </div>`;
  }
}
