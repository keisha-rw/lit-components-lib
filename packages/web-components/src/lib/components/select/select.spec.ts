import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, elementUpdated, fixture } from '@open-wc/testing';
import userEvent from '@testing-library/user-event';
import { PdsSelect } from './select';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Select/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('Select unit tests', () => {
  let element: PdsSelect;
  beforeEach(async () => {
    element = await fixture(`<pds-select
    label="select label"
    variant="default"
    name="select name"
    placeholder="select placeholder"
  >
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
    <option value="3">Option 3</option>
  </pds-select>`);
  });

  it('is an instance of PdsSelect', () => {
    expect(element).toBeInstanceOf(PdsSelect);
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('dispatches a pds-select-blur event on blur', async () => {
    jest.spyOn(element, 'dispatchEvent');
    const userEventWithoutDelay = userEvent.setup({ delay: null });

    const selectElement = element.shadowRoot?.querySelector('select');
    if (selectElement) {
      await userEventWithoutDelay.click(selectElement);
      await userEventWithoutDelay.tab();
    }
    const blurEvent = new Event('pds-select-blur', {
      bubbles: false,
      composed: true,
    });
    expect(element.dispatchEvent).toHaveBeenCalledWith(blurEvent);
  });

  it('dispatches a pds-select-change event on change', async () => {
    jest.spyOn(element, 'dispatchEvent');
    const userEventWithoutDelay = userEvent.setup({ delay: null });

    const selectElement = element.shadowRoot?.querySelector('select');
    if (selectElement) {
      await userEventWithoutDelay.selectOptions(selectElement, ['1']);
      // Change the value so we can verify that the onChange was fired
      await userEventWithoutDelay.selectOptions(selectElement, ['2']);
    }
    expect(element.value).toStrictEqual('2');
    const changeEvent = new Event('pds-select-change', {
      bubbles: false,
      composed: true,
    });
    expect(element.dispatchEvent).toHaveBeenCalledWith(changeEvent);
  });

  it('returns nothing and logs an error if there is no label', async () => {
    // Mock the console error so it doesn't muck up our testing logs
    jest.spyOn(console, 'error').mockImplementation(() => {});
    await fixture(`<pds-select
    variant="default"
    name="select name"
    placeholder="select placeholder"
  >
    <option value="1">Option 1</option>
    <option value="2">Option 2</option>
    <option value="3">Option 3</option>
  </pds-select>`);

    expect(console.error).toHaveBeenCalledTimes(1);
    expect(console.error).toHaveBeenCalledWith(
      '"label" is required as a property or as a slot in <%s /> but is undefined',
      'pds-select',
    );
  });

  it('sets selected on the correct option when defaultValue is set', async () => {
    const elementWithDefaultValue = await fixture(`<pds-select
    variant="default"
    label="select label"
    name="select name"
    value="option2"
  >
    <option value="option1">Option 1</option>
    <option value="option2">Option 2</option>
    <option value="option3">Option 3</option>
  </pds-select>`);

    // Grab option 2
    const selectedOption = elementWithDefaultValue.shadowRoot?.querySelector(
      'option[value="option2"]',
    );

    // Make sure it's selected
    expect(selectedOption).toMatchSnapshot();
  });

  it('renders a placeholder when placeholder attribute is set', () => {
    // Query placeholder by ID
    const placeholder = element.shadowRoot?.querySelector(
      'option[id="placeholder"]',
    );

    // Make sure placeholder is set
    expect(placeholder?.textContent).toBe('select placeholder');
  });

  it('does not render a placeholder when placeholder attribute is not set', async () => {
    const elementWithoutPlaceholder = await fixture(`<pds-select
      variant="default"
      label="select label"
      name="select name"
    >
      <option value="option1">Option 1</option>
      <option value="option2">Option 2</option>
      <option value="option3">Option 3</option>
    </pds-select>`);

    // Query placeholder by ID
    const emptyPlaceholder =
      elementWithoutPlaceholder.shadowRoot?.querySelector(
        'option[id="placeholder"]',
      );

    // Make sure placeholder is null
    expect(emptyPlaceholder).toBeNull();
  });

  it('should reset the select value on form reset', async () => {
    const originalValue = 'option2';
    const newValue = 'option3';
    const form: HTMLFormElement = await fixture(
      `<form><pds-select
      variant="default"
      label="select label"
      name="select name"
      value="${originalValue}"
    >
      <option value="option1">Option 1</option>
      <option value="option2">Option 2</option>
      <option value="option3">Option 3</option>
    </pds-select></form>`,
    );

    element = form.querySelector('pds-select') as PdsSelect;

    const selectEl = element?.shadowRoot?.querySelector(
      'select',
    ) as HTMLSelectElement;

    element.value = newValue;

    expect(element.value).toBe(newValue);
    await elementUpdated(element);
    expect(selectEl.selectedOptions[0].value).toBe(newValue);

    form.reset();

    expect(element.value).toBe(originalValue);
    await elementUpdated(element);
    expect(selectEl.selectedOptions[0].value).toBe(originalValue);
  });

  it('selectContainsOption returns correct boolean for option existence in the select', async () => {
    expect(element.selectContainsOption(element.field, '1')).toBeTruthy();
    expect(element.selectContainsOption(element.field, '5')).toBeFalsy();
  });

  it('handles deletion of an option attribute', async () => {
    const elementWithDefaultValue = await fixture(`<pds-select
    variant="default"
    label="select label"
    name="select name"
    value="option1"
  >
    <option value="option1">Option 1</option>
    <option value="option2">Option 2</option>
    <option value="option3">Option 3</option>
  </pds-select>`);
    elementWithDefaultValue.querySelector('option')?.remove();

    await new Promise(process.nextTick);
    await (elementWithDefaultValue as PdsSelect).updateComplete;
    expect((elementWithDefaultValue as PdsSelect).value).toBe('option2');
    expect((elementWithDefaultValue as PdsSelect).field.options).toHaveLength(
      2,
    );
  });

  it('handles modification of an option attribute', async () => {
    const elementWithDefaultValue = await fixture(`<pds-select
    variant="default"
    label="select label"
    name="select name"
    value="option2"
  >
    <option value="option1">Option 1</option>
    <option value="option2">Option 2</option>
    <option value="option3">Option 3</option>
  </pds-select>`);

    elementWithDefaultValue.querySelector('option')?.setAttribute('value', '1');

    await new Promise(process.nextTick);
    await (elementWithDefaultValue as PdsSelect).updateComplete;
    expect((elementWithDefaultValue as PdsSelect).value).toBe('option2');
  });

  it('handles modification of an option text node', async () => {
    const elementWithDefaultValue = await fixture(`<pds-select
    variant="default"
    label="select label"
    name="select name"
    value="option2"
  >
    <option value="option1">Option 1</option>
    <option value="option2">Option 2</option>
    <option value="option3">Option 3</option>
  </pds-select>`);

    // @ts-expect-error ts thinks this could be null
    elementWithDefaultValue.querySelector('option').text = 'blast';

    await new Promise(process.nextTick);
    await (elementWithDefaultValue as PdsSelect).updateComplete;
    expect((elementWithDefaultValue as PdsSelect).value).toBe('option2');
  });

  it('handles addition of an option attribute', async () => {
    const elementWithDefaultValue = await fixture(`<pds-select
    variant="default"
    label="select label"
    name="select name"
    value="option2"
  >
    <option value="option1">Option 1</option>
    <option value="option2">Option 2</option>
    <option value="option3">Option 3</option>
  </pds-select>`);

    elementWithDefaultValue.insertAdjacentHTML(
      'beforeend',
      '<option value="option4">Option 4</option>',
    );

    await new Promise(process.nextTick);
    expect((elementWithDefaultValue as PdsSelect).value).toBe('option2');
    expect((elementWithDefaultValue as PdsSelect).field.options).toHaveLength(
      4,
    );
  });
});
