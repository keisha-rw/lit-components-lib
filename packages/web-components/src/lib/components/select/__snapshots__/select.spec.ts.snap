// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select unit tests sets selected on the correct option when defaultValue is set 1`] = `
<option
  value="option2"
>
  Option 2
</option>
`;
